src/main/java/com/cloudant/search3/SearchServer.java

```diff
@@ -0,0 +1,273 @@
+            // cribbed from ProtobufDecoder.decode().
+            final ByteBuf msg = req.content();
+            final byte[] array;
+            final int offset;
+            final int length = msg.readableBytes();
+            if (msg.hasArray()) {
+                array = msg.array();
+                offset = msg.arrayOffset() + msg.readerIndex();
+            } else {
+                array = ByteBufUtil.getBytes(msg, msg.readerIndex(), length, false);
+                offset = 0;
+            }
+            return (T) prototype.getParserForType().parseFrom(array, offset, length);
+        }
+
+        private ByteBuf encode(final MessageLite msg) {
+            return wrappedBuffer(msg.toByteArray());
+        }
+
+        private FullHttpResponse error(final Type type, final String reason) {
+            final ErrorResponse error = ErrorResponse.newBuilder().setType(type).setReason(reason).build();
+            final ByteBuf body = wrappedBuffer(error.toByteArray());
+            final FullHttpResponse result = new DefaultFullHttpResponse(
+                    HttpVersion.HTTP_1_1,
+                    OK,
+                    body);
+            result.headers()
+            .set(RPC_MESSAGE_TYPE, "error_response")
+            .set(CONTENT_TYPE, APPLICATION_OCTET_STREAM)
+            .setInt(CONTENT_LENGTH, result.content().readableBytes());
+            return result;
+        }
+
+    }
+
+    private static class SearchServerInitializer extends ChannelInitializer<SocketChannel> {
+
+        private static final EventExecutorGroup searchGroup = new DefaultEventExecutorGroup(16);
+
+        private final SslContext sslCtx;
+        private final Search search;
+
+        public SearchServerInitializer(final SslContext sslCtx, final Search search) {
+            this.sslCtx = sslCtx;
+            this.search = search;
+        }
+
+        @Override
+        public void initChannel(SocketChannel ch) {
+            final ChannelPipeline p = ch.pipeline();
+            if (sslCtx != null) {
+                p.addLast(sslCtx.newHandler(ch.alloc()));
+            }
+
+            // http
+            p.addLast("httpCodec", new HttpServerCodec());
+            p.addLast("httpAggregator", new HttpObjectAggregator(10485760));
+
+            // inbound
+            p.addLast("httpExpectContinue", new HttpServerExpectContinueHandler());
+            p.addLast("httpKeepAlive", new HttpServerKeepAliveHandler());
+            p.addLast("httpDecompressor", new HttpContentDecompressor());
+
+            // outbound
+            p.addLast("httpCompressor", new HttpContentCompressor());
+
+            // Search handler in a separate group to the I/O.
+            p.addLast(searchGroup, "searchHandler", new SearchServerHandler(search));
+        }
+
+    }
+
+    public SearchServer(final int port, final SslContext sslCtx, final Search search) {
```

Comment:
is it customary to put the constructor all the way down here behind the private classes?

Reply:
we haven't picked a Java house style but we could. I don't particularly care about the ordering, happy to change it if we agree on a style.

Reply:
I am +1 to a style discussion and to apply a style to our Java components.